src/tests/dotnet-counters/CounterMonitorPayloadTests.cs

```diff
@@ -0,0 +1,311 @@
+using System.Linq;
+using System.Text.Json;
+using System.Threading;
+using System.Threading.Tasks;
+using CommonTestRunner;
+using Microsoft.Diagnostics.TestHelpers;
+using Microsoft.Diagnostics.Tools.Counters;
+using Xunit;
+using Xunit.Abstractions;
+using Xunit.Extensions;
+using TestRunner = Microsoft.Diagnostics.CommonTestRunner.TestRunner;
+using Constants = DotnetCounters.UnitTests.TestConstants;
+
+namespace DotnetCounters.UnitTests
+{
+    /// <summary>
+    /// Tests the behavior of CounterMonitor's Collect command.
+    /// </summary>
+    public class CounterMonitorPayloadTests
+    {
+        private ITestOutputHelper _outputHelper;
+        private static readonly TimeSpan DefaultTimeout = TimeSpan.FromMinutes(2);
+        private static readonly string SystemRuntimeName = "System.Runtime";
+        private static readonly string Metric = "Metric";
+        private static readonly string Rate = "Rate";
+        private static readonly string TagStart = "[";
+
+        private string[] ExpectedCounterTypes = { Metric, Rate };
+
+        public CounterMonitorPayloadTests(ITestOutputHelper outputHelper)
+        {
+            _outputHelper = outputHelper;
+        }
+
+        [SkippableTheory, MemberData(nameof(Configurations))]
+        public async Task TestCounterMonitorCustomMetricsJSON(TestConfiguration configuration)
+        {
+            CheckFramework(configuration.BuildProjectFramework);
+
+            JSONCounterTrace trace = await GetCounterTraceJSON(configuration, new List<string> { Constants.TestMeterName });
+            Assert.NotEmpty(trace.events);
+
+            ValidateCustomMetrics(
+                trace.events.Select(e => e.provider).ToHashSet(),
+                trace.events.Select(e => e.name).ToHashSet(),
+                trace.events.Select(e => e.counterType).Distinct(),
+                trace.events.Where(e => e.name.Equals(Constants.TestHistogramName)).Select(e => e.tags).Where(t => !string.IsNullOrEmpty(t)).ToHashSet(),
+                trace.events.Where(e => e.name.Equals(Constants.TestCounterName)).Select(e => e.tags).Where(t => !string.IsNullOrEmpty(t)).ToHashSet(),
+                trace.events.Where(e => e.name.Equals(Constants.TestHistogramName)).Select(e => e.value).ToList(),
+                trace.events.Where(e => e.name.Equals(Constants.TestCounterName)).Select(e => e.value).ToList(),
+                CountersExportFormat.json
+                );
+        }
+
+        [SkippableTheory, MemberData(nameof(Configurations))]
+        public async Task TestCounterMonitorCustomMetricsCSV(TestConfiguration configuration)
+        {
+            CheckFramework(configuration.BuildProjectFramework);
+
+            List<string> lines = await GetCounterTraceCSV(configuration, new List<string> { Constants.TestMeterName });
+            IEnumerable<string[]> splitLines = lines.Select(l => l.Split(","));
+
+            var providers = splitLines.Select(line => line[Constants.ProviderIndex]).ToHashSet();
```

Comment:
nit: Let's assert a minimum length that we expect

Reply:
Just to keep the parsing separate from the assertions, the validation happens in `ValidateCustomMetrics` - I could also add it here, but it would be redundant